File: task3.py
import aiohttp
import asyncio
import json
from typing import List, Dict, Optional
import sys
import os
from tabulate import tabulate  # For table formatting
import re  # For regular expressions

def clear_screen() -> None:
    """Clear the terminal screen for both Windows and Unix-like systems"""
    os.system('cls' if os.name == 'nt' else 'clear')

async def fetch_users() -> Optional[List[Dict]]:
    """
    Asynchronously fetch users from the API with comprehensive error handling
    """
    url = "https://jsonplaceholder.typicode.com/users"
    timeout = aiohttp.ClientTimeout(total=10)  # 10 seconds timeout
    
    try:
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.get(url) as response:
                # Check if the response status is OK (200)
                response.raise_for_status()
                
                # Try to parse JSON response
                return await response.json()
                
    except aiohttp.ClientTimeout:
        print("\nError: The request timed out. Please check your internet connection and try again.")
    except aiohttp.ClientConnectionError:
        print("\nError: Could not connect to the server. Please check your internet connection.")
    except aiohttp.ContentTypeError:
        print("\nError: Received invalid JSON response from the server.")
    except aiohttp.ClientResponseError as e:
        if e.status == 404:
            print("\nError: The requested resource was not found.")
        elif e.status == 403:
            print("\nError: Access to the resource is forbidden.")
        else:
            print(f"\nError: Server returned an error (Status code: {e.status})")
    except Exception as e:
        print(f"\nAn unexpected error occurred: {str(e)}")
    
    return None

def filter_users_by_name(users: List[Dict], search_term: str) -> List[Dict]:
    """
    Filter users based on regular expression pattern (case-insensitive)
    Examples:
        - "^A" will match names starting with A
        - "a$" will match names ending with a
        - ".*son.*" will match names containing "son"
        - "[AM].*" will match names starting with A or M
    """
    if not users:
        return []
        
    # Handle empty search term
    if not search_term or search_term.isspace():
        print("\nWarning: Empty search term - returning all users")
        return users
    
    try:
        # Try to compile the regular expression pattern
        try:
            # Escape special characters if the pattern starts with a backslash
            if search_term.startswith('\\'):
                search_term = re.escape(search_term[1:])
                print(f"\nUsing literal search pattern: {search_term}")
            else:
                print(f"\nUsing regex pattern: {search_term}")
            
            pattern = re.compile(search_term, re.IGNORECASE)
            
        except re.error as e:
            print(f"\nInvalid regular expression: {str(e)}")
            print("Tips:")
            print("- Use \\. to match a literal dot")
            print("- Use \\* to match a literal asterisk")
            print("- Use \\[ to match a literal square bracket")
            print("- Add \\ before any special character to match it literally")
            print("\nFalling back to plain text search...")
            # Fallback to plain text search if regex is invalid
            pattern = re.compile(re.escape(search_term), re.IGNORECASE)
        
        # Perform the search with detailed error handling
        try:
            filtered_users = []
            print("\nDebug: Matching names:")
            for user in users:
                name = user['name']
                if pattern.search(name):
                    filtered_users.append(user)
                    print(f"  ✓ '{name}' matches pattern")
                else:
                    print(f"  × '{name}' does not match pattern")
            
            # Provide feedback about the search results
            if not filtered_users:
                print(f"\nNo users found matching pattern '{search_term}' in their name.")
            else:
                print(f"\nFound {len(filtered_users)} user(s) matching pattern '{search_term}' in their name:")
                for user in filtered_users:
                    print(f"  - {user['name']}")
                
            return filtered_users
            
        except Exception as search_error:
            print(f"\nError during pattern matching: {str(search_error)}")
            print("Falling back to plain text search...")
            # Ultimate fallback - plain text contains search
            return [user for user in users if search_term.lower() in user['name'].lower()]
            
    except Exception as e:
        print(f"\nUnexpected error while filtering users: {str(e)}")
        print("Returning all users...")
        return users

def display_users(users: Optional[List[Dict]]) -> None:
    """
    Display user information with error handling
    """
    clear_screen()
    if not users:
        print("\nNo users found matching the search criteria.")
        return
    
    try:
        for user in users:
            print(f"\nName: {user['name']}")
            print(f"Email: {user['email']}")
            print(f"Address: {user['address']['street']}, {user['address']['city']}")
            print("-" * 50)
        
        # Add pause at the end of display
        input("\nPress Enter to continue...")
        clear_screen()
    except KeyError as e:
        print(f"\nError: Missing required field in user data: {str(e)}")
    except Exception as e:
        print(f"\nError while displaying users: {str(e)}")

def display_users_json(users: List[Dict]) -> None:
    """Display users in JSON format"""
    clear_screen()
    try:
        formatted_json = json.dumps(users, indent=2)
        print("\nUsers in JSON format:")
        print(formatted_json)
        input("\nPress Enter to continue...")
        clear_screen()
    except Exception as e:
        print(f"\nError formatting JSON: {str(e)}")

def display_users_table(users: List[Dict]) -> None:
    """Display users in table format"""
    clear_screen()
    try:
        # Prepare data for table
        table_data = []
        headers = ["Name", "Email", "Street", "City"]
        
        for user in users:
            row = [
                user['name'],
                user['email'],
                user['address']['street'],
                user['address']['city']
            ]
            table_data.append(row)
        
        print("\nUsers in table format:")
        print(tabulate(table_data, headers=headers, tablefmt="grid"))
        input("\nPress Enter to continue...")
        clear_screen()
    except Exception as e:
        print(f"\nError creating table: {str(e)}")

def display_users_compact(users: List[Dict]) -> None:
    """Display users in compact format"""
    clear_screen()
    try:
        print("\nUsers in compact format:")
        for user in users:
            print(f"{user['name']} | {user['email']} | {user['address']['street']}, {user['address']['city']}")
        input("\nPress Enter to continue...")
        clear_screen()
    except Exception as e:
        print(f"\nError displaying compact format: {str(e)}")

def choose_display_format(users: List[Dict]) -> None:
    """Choose and apply display format for users"""
    while True:
        clear_screen()
        print("\nChoose display format:")
        print("1. Standard format")
        print("2. JSON format")
        print("3. Table format")
        print("4. Compact format")
        print("5. Back to main menu")
        
        choice = input("\nEnter your choice (1-5): ").strip()
        
        if choice == "1":
            display_users(users)
        elif choice == "2":
            display_users_json(users)
        elif choice == "3":
            display_users_table(users)
        elif choice == "4":
            display_users_compact(users)
        elif choice == "5":
            clear_screen()
            break
        else:
            print("\nInvalid choice. Please enter 1-5.")
            input("\nPress Enter to continue...")

async def main() -> None:
    """
    Main program loop with error handling
    """
    # Fetch all users
    clear_screen()
    print("\nFetching users from the API...")
    users = await fetch_users()
    
    if users is None:
        print("\nCould not fetch user data. Please try again later.")
        return
    
    while True:
        try:
            #clear_screen()
            print("\nOptions:")
            print("1. Show all users")
            print("2. Filter users by name (supports regex)")
            print("3. Change display format")
            print("4. Exit")
            
            choice = input("\nEnter your choice (1-4): ").strip()
            
            
            if choice == "1":
                print("\nShowing all users:")
                display_users(users)
            
            elif choice == "2":
                clear_screen()
                print("\nRegex Pattern Examples:")
                print("  ^A        - names starting with A")
                print("  a$        - names ending with a")
                print("  .*son.*   - names containing 'son'")
                print("  [AM].*    - names starting with A or M")
                print("  \\bJohn\\b - match whole word 'John'")
                print("\nSpecial Characters:")
                print("  Add \\ before . * + ? ^ $ [ ] ( ) { } | \\ to match them literally")
                print("  Example: \\. matches literal dot, \\* matches literal asterisk")
                print("\nAvailable names for reference:")
                for user in users:
                    print(f"  - {user['name']}")
                
                search_term = input("\nEnter a regex pattern to search for (or '\\' + text for literal search): ").strip()
                if not search_term:
                    print("\nEmpty search term - showing all users")
                    display_users(users)
                    continue
                    
                #clear_screen()
                filtered_users = filter_users_by_name(users, search_term)
                if filtered_users:  # Only show users if we found any
                    print(f"\nShowing users matching pattern '{search_term}':")
                    display_users(filtered_users)
            
            elif choice == "3":
                choose_display_format(users)
            
            elif choice == "4":
                clear_screen()
                print("\nGoodbye!")
                break
            
            else:
                print("\nInvalid choice. Please enter 1-4.")
                input("\nPress Enter to continue...")
                
        except KeyboardInterrupt:
            clear_screen()
            print("\n\nProgram interrupted by user. Exiting...")
            break
        except Exception as e:
            print(f"\nAn unexpected error occurred: {str(e)}")
            print("Please try again.")
            input("\nPress Enter to continue...")

if __name__ == "__main__":
    try:
        if sys.platform == 'win32':
            # Set up proper event loop policy for Windows
            asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
        
        # Run the async main function
        asyncio.run(main())
    except KeyboardInterrupt:
        clear_screen()
        print("\nProgram terminated by user.")
    except Exception as e:
        print(f"\nFatal error: {str(e)}")

File: TS/src/task3.ts
import axios from 'axios';
import { table } from 'table';
import * as os from 'os';

// Types
interface Address {
  street: string;
  city: string;
  zipcode: string;
  suite: string;
  geo: {
    lat: string;
    lng: string;
  };
}

interface User {
  id: number;
  name: string;
  email: string;
  address: Address;
  phone: string;
  website: string;
  company: {
    name: string;
    catchPhrase: string;
    bs: string;
  };
}

// Utility functions
function clearScreen(): void {
  console.clear();
}

async function fetchUsers(): Promise<User[] | null> {
  const url = "https://jsonplaceholder.typicode.com/users";
  const timeout = 10000; // 10 seconds

  try {
    console.log("\nFetching users from the API...");
    const response = await axios.get<User[]>(url, { timeout });
    
    // Validate user data
    console.log("\nValidating user data...");
    const validUsers = response.data.filter(user => {
      if (typeof user === 'object' && user.name && typeof user.name === 'string') {
        console.log(`  ✓ Valid user: ${user.name}`);
        return true;
      }
      console.log(`  × Invalid user data: ${JSON.stringify(user)}`);
      return false;
    });

    if (validUsers.length === 0) {
      console.log("\nError: No valid users found in API response");
      return null;
    }

    console.log(`\nSuccessfully loaded ${validUsers.length} valid users`);
    return validUsers;

  } catch (error) {
    if (axios.isAxiosError(error)) {
      if (error.code === 'ECONNABORTED') {
        console.log("\nError: The request timed out. Please check your internet connection and try again.");
      } else if (error.response) {
        if (error.response.status === 404) {
          console.log("\nError: The requested resource was not found.");
        } else if (error.response.status === 403) {
          console.log("\nError: Access to the resource is forbidden.");
        } else {
          console.log(`\nError: Server returned an error (Status code: ${error.response.status})`);
        }
      } else if (error.request) {
        console.log("\nError: Could not connect to the server. Please check your internet connection.");
      }
    } else {
      console.log(`\nAn unexpected error occurred: ${error}`);
    }
    return null;
  }
}

function filterUsersByName(users: User[], searchTerm: string): User[] {
  if (!users) {
    return [];
  }

  if (!searchTerm || searchTerm.trim() === '') {
    console.log("\nWarning: Empty search term - returning all users");
    return users;
  }

  try {
    // Print available names for debugging
    console.log("\nAvailable names in database:");
    users.forEach(user => console.log(`  • ${user.name}`));

    let pattern: RegExp;
    try {
      if (searchTerm.startsWith('\\')) {
        const literalSearch = searchTerm.slice(1);
        pattern = new RegExp(literalSearch.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'i');
        console.log(`\nUsing literal search pattern: ${literalSearch}`);
      } else {
        pattern = new RegExp(searchTerm, 'i');
        console.log(`\nUsing regex pattern: ${searchTerm}`);
      }
    } catch (e) {
      console.log(`\nInvalid regular expression: ${e}`);
      console.log("Tips:");
      console.log("- Use \\. to match a literal dot");
      console.log("- Use \\* to match a literal asterisk");
      console.log("- Use \\[ to match a literal square bracket");
      console.log("- Add \\ before any special character to match it literally");
      console.log("\nFalling back to plain text search...");
      pattern = new RegExp(searchTerm.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'i');
    }

    const filteredUsers = users.filter(user => {
      const match = pattern.exec(user.name);
      if (match) {
        console.log(`  ✓ '${user.name}' matches at position ${match.index}-${match.index + match[0].length}: '${match[0]}'`);
        return true;
      }
      console.log(`  × '${user.name}' does not match`);
      return false;
    });

    if (filteredUsers.length === 0) {
      console.log(`\nNo users found matching pattern '${searchTerm}' in their name.`);
    } else {
      console.log(`\nFound ${filteredUsers.length} user(s) matching pattern '${searchTerm}' in their name:`);
      filteredUsers.forEach(user => console.log(`  - ${user.name}`));
    }

    return filteredUsers;

  } catch (error) {
    console.log(`\nUnexpected error while filtering users: ${error}`);
    console.log("Returning all users...");
    return users;
  }
}

function displayUsers(users: User[] | null): void {
  clearScreen();
  if (!users || users.length === 0) {
    console.log("\nNo users found matching the search criteria.");
    return;
  }

  try {
    users.forEach(user => {
      console.log(`\nName: ${user.name}`);
      console.log(`Email: ${user.email}`);
      console.log(`Address: ${user.address.street}, ${user.address.city}`);
      console.log("-".repeat(50));
    });

    // Add pause at the end of display
    console.log("\nPress Enter to continue...");
    require('readline').createInterface({
      input: process.stdin,
      output: process.stdout
    }).question('', () => {
      clearScreen();
      process.stdin.resume();
    });
  } catch (error) {
    console.log(`\nError while displaying users: ${error}`);
  }
}

function displayUsersJson(users: User[]): void {
  clearScreen();
  try {
    const formattedJson = JSON.stringify(users, null, 2);
    console.log("\nUsers in JSON format:");
    console.log(formattedJson);
    console.log("\nPress Enter to continue...");
    require('readline').createInterface({
      input: process.stdin,
      output: process.stdout
    }).question('', () => {
      clearScreen();
      process.stdin.resume();
    });
  } catch (error) {
    console.log(`\nError formatting JSON: ${error}`);
  }
}

function displayUsersTable(users: User[]): void {
  clearScreen();
  try {
    const headers = ["Name", "Email", "Street", "City"];
    const data = [
      headers,
      ...users.map(user => [
        user.name,
        user.email,
        user.address.street,
        user.address.city
      ])
    ];

    console.log("\nUsers in table format:");
    console.log(table(data));
    console.log("\nPress Enter to continue...");
    require('readline').createInterface({
      input: process.stdin,
      output: process.stdout
    }).question('', () => {
      clearScreen();
      process.stdin.resume();
    });
  } catch (error) {
    console.log(`\nError creating table: ${error}`);
  }
}

function displayUsersCompact(users: User[]): void {
  clearScreen();
  try {
    console.log("\nUsers in compact format:");
    users.forEach(user => {
      console.log(`${user.name} | ${user.email} | ${user.address.street}, ${user.address.city}`);
    });
    console.log("\nPress Enter to continue...");
    require('readline').createInterface({
      input: process.stdin,
      output: process.stdout
    }).question('', () => {
      clearScreen();
      process.stdin.resume();
    });
  } catch (error) {
    console.log(`\nError displaying compact format: ${error}`);
  }
}

async function chooseDisplayFormat(users: User[]): Promise<void> {
  while (true) {
    clearScreen();
    console.log("\nChoose display format:");
    console.log("1. Standard format");
    console.log("2. JSON format");
    console.log("3. Table format");
    console.log("4. Compact format");
    console.log("5. Back to main menu");

    const rl = require('readline').createInterface({
      input: process.stdin,
      output: process.stdout
    });

    const choice = await new Promise<string>(resolve => {
      rl.question('\nEnter your choice (1-5): ', resolve);
    });
    rl.close();

    switch (choice.trim()) {
      case "1":
        displayUsers(users);
        break;
      case "2":
        displayUsersJson(users);
        break;
      case "3":
        displayUsersTable(users);
        break;
      case "4":
        displayUsersCompact(users);
        break;
      case "5":
        clearScreen();
        return;
      default:
        console.log("\nInvalid choice. Please enter 1-5.");
        console.log("\nPress Enter to continue...");
        await new Promise(resolve => {
          require('readline').createInterface({
            input: process.stdin,
            output: process.stdout
          }).question('', resolve);
        });
    }
  }
}

async function main(): Promise<void> {
  // Fetch all users
  clearScreen();
  const users = await fetchUsers();

  if (!users) {
    console.log("\nCould not fetch user data. Please try again later.");
    return;
  }

  while (true) {
    try {
      console.log("\nOptions:");
      console.log("1. Show all users");
      console.log("2. Filter users by name (supports regex)");
      console.log("3. Change display format");
      console.log("4. Exit");

      const rl = require('readline').createInterface({
        input: process.stdin,
        output: process.stdout
      });

      const choice = await new Promise<string>(resolve => {
        rl.question('\nEnter your choice (1-4): ', resolve);
      });
      rl.close();

      switch (choice.trim()) {
        case "1":
          console.log("\nShowing all users:");
          displayUsers(users);
          break;

        case "2":
          clearScreen();
          console.log("\nRegex Pattern Examples:");
          console.log("  ^A        - names starting with A");
          console.log("  a$        - names ending with a");
          console.log("  .*son.*   - names containing 'son'");
          console.log("  [AM].*    - names starting with A or M");
          console.log("  \\bJohn\\b - match whole word 'John'");
          console.log("\nSpecial Characters:");
          console.log("  Add \\ before . * + ? ^ $ [ ] ( ) { } | \\ to match them literally");
          console.log("  Example: \\. matches literal dot, \\* matches literal asterisk");
          console.log("\nAvailable names for reference:");
          users.forEach(user => console.log(`  - ${user.name}`));

          const searchTerm = await new Promise<string>(resolve => {
            require('readline').createInterface({
              input: process.stdin,
              output: process.stdout
            }).question('\nEnter a regex pattern to search for (or \'\\\' followed by text for literal search): ', resolve);
          });

          if (!searchTerm) {
            console.log("\nEmpty search term - showing all users");
            displayUsers(users);
            continue;
          }

          const filteredUsers = filterUsersByName(users, searchTerm);
          if (filteredUsers.length > 0) {
            console.log(`\nShowing users matching pattern '${searchTerm}':`);
            displayUsers(filteredUsers);
          }
          break;

        case "3":
          await chooseDisplayFormat(users);
          break;

        case "4":
          clearScreen();
          console.log("\nGoodbye!");
          return;

        default:
          console.log("\nInvalid choice. Please enter 1-4.");
          console.log("\nPress Enter to continue...");
          await new Promise(resolve => {
            require('readline').createInterface({
              input: process.stdin,
              output: process.stdout
            }).question('', resolve);
          });
      }
    } catch (error) {
      if (error instanceof Error && error.name === 'SyntaxError') {
        console.log("\nInvalid input. Please try again.");
      } else {
        console.log(`\nAn unexpected error occurred: ${error}`);
        console.log("Please try again.");
      }
      console.log("\nPress Enter to continue...");
      await new Promise(resolve => {
        require('readline').createInterface({
          input: process.stdin,
          output: process.stdout
        }).question('', resolve);
      });
    }
  }
}

// Run the program
if (require.main === module) {
  main().catch(error => {
    console.log(`\nFatal error: ${error}`);
    process.exit(1);
  });
} 